lib/main.ts

```diff
@@ -52,636 +52,565 @@ const SECRETS = JSON.parse(shell('sh ./generate_secrets.sh').toString());
-        message.content = trimmed;
-        // When finished expanding...
-        this.expand(message).then(content => {
-            if (content.length >= 2000) {
-                message.answer("The expansion for that message was"
-                + " over 2000 characters, what the fuck is wrong with you?");
-                return;
-            }
-            message.content = content;
-            console.log('Expanded message:', message.content);
-
-            if (message.content[0] === CONFIG.commands.prefix) {
-                console.log('Message type: command.')
-                this.process_command(message);
-            } else {
-                console.log('Message type: generic.')
-                this.process_generic(message);
-            }
-        });
-    }
+	private static _client : Client;
+	private _COMMAND_HISTORY : Message[] = [];
+
+	constructor() {
+		console.log('Secrets:', pp(SECRETS));
+		console.log('Configured Variables:', pp(CONFIG));
+		console.log('Known commands:', pp(KNOWN_COMMANDS));
+	}
+
+	static start() {
+		this._client = new Client();
+		this._client.login(
+			SECRETS.api.token,
+			`${__dirname}/*Discord.ts`
+		);
+	}
+
+	process_command(message : Message) {
+		const last_command = this._COMMAND_HISTORY.last();
+		this._COMMAND_HISTORY.push(message);
+		if (this._COMMAND_HISTORY.length > CONFIG.commands.max_history) {
+			this._COMMAND_HISTORY.shift();
+		}
+		const current_command = this._COMMAND_HISTORY.last();
+
+		// Try and slow the fellas down a little.
+		if (last_command.channel === current_command.channel) {
+			// Only give spam warning if commands are coming
+			//  fast _in the same channel_.
+			const delta = current_command.createdTimestamp - last_command.createdTimestamp;
+			if (last_command.content === current_command.content
+				&& delta <= 1400) {
+				if (delta <= 400) return;
+				return message.answer(`I can't help but notice you're running \
+					the same commands over in rather rapid succession.
+					Would you like to slow down a little?`.squeeze())
+			}
+			if (delta <= 900) {
+				if (delta <= 300) return;
+				return message.answer('Slow down there bucko.');
+			}
+		}
+
+		const content = message.content.trim().squeeze();
+		const words = content.tail().split(' ');
+		const args = words.tail();
+
+		let command = words[0].toLowerCase();
+		if (CONFIG.commands.aliases.hasOwnProperty(command))
+			command = CONFIG.commands.aliases[command].trim().squeeze();
+
+		const expanded_command_words = command.split(' ');
+		if (expanded_command_words.length > 1) {
+			// This means the alias has expanded to more than just one word.
+			command = expanded_command_words.shift();
+			expanded_command_words.each(e => args.unshift(e));
+		}
+
+		command = command.toLowerCase();
+		console.log('Received command:', [command, args]);
+
+		var functions = readdirSync('funcs').map(n => n.slice(0, -3))
+		require('./funcs/' + functions[command] + '.js')(message, args, SECRETS, CONFIG)
```

Comment:
That's going to fail on most commands, you should check if the file exists first, no?

Reply:
```js
if (command in functions) {
	// Then require.
}
```

Reply:
functions[command] || 'fallback' innit